File: mobile/flutter/lib/api/resilience.dart
import 'dart:async';

/// Thrown instead of calling the API while the breaker is open.
class CircuitOpenException implements Exception {
  CircuitOpenException(this.retryAfter);
  final Duration retryAfter;
  @override
  String toString() => 'CircuitOpenException(retry in ${retryAfter.inSeconds}s)';
}

enum BreakerState { closed, open, halfOpen }

/// Client-side circuit breaker. A phone that keeps retrying a dead API drains
/// the battery and delays the moment the UI can honestly say "offline", so
/// failing fast is a UX decision here, not just an infrastructure one.
class CircuitBreaker {
  CircuitBreaker({
    this.failureThreshold = 4,
    this.openDuration = const Duration(seconds: 20),
    DateTime Function()? now,
  }) : _now = now ?? DateTime.now;

  final int failureThreshold;
  final Duration openDuration;
  final DateTime Function() _now;

  BreakerState _state = BreakerState.closed;
  int _failures = 0;
  DateTime? _openedAt;

  BreakerState get state {
    _refresh();
    return _state;
  }

  Future<T> run<T>(Future<T> Function() action) async {
    _refresh();
    if (_state == BreakerState.open) {
      final elapsed = _now().difference(_openedAt!);
      throw CircuitOpenException(openDuration - elapsed);
    }
    try {
      final result = await action();
      _onSuccess();
      return result;
    } catch (_) {
      _onFailure();
      rethrow;
    }
  }

  void _refresh() {
    if (_state == BreakerState.open &&
        _now().difference(_openedAt!) >= openDuration) {
      _state = BreakerState.halfOpen;
    }
  }

  void _onSuccess() {
    _state = BreakerState.closed;
    _failures = 0;
  }

  void _onFailure() {
    if (_state == BreakerState.halfOpen) {
      _state = BreakerState.open;
      _openedAt = _now();
      return;
    }
    _failures++;
    if (_failures >= failureThreshold) {
      _state = BreakerState.open;
      _openedAt = _now();
    }
  }
}

/// Token-bucket throttle for outbound calls: it keeps a chatty screen (or a
/// rebuild loop) from turning into a request storm the server has to absorb.
class RateLimiter {
  RateLimiter({this.perSecond = 5, this.burst = 10, DateTime Function()? now})
      : _tokens = burst.toDouble(),
        _now = now ?? DateTime.now;

  final double perSecond;
  final int burst;
  final DateTime Function() _now;

  double _tokens;
  DateTime? _last;

  bool tryAcquire() {
    if (perSecond <= 0) return true;
    final now = _now();
    _last ??= now;
    _tokens += now.difference(_last!).inMicroseconds / 1e6 * perSecond;
    if (_tokens > burst) _tokens = burst.toDouble();
    _last = now;
    if (_tokens < 1) return false;
    _tokens -= 1;
    return true;
  }
}

File: mobile/ios-native/Sources/TaskTrooperIOS/Data/Resilience.swift
import Foundation

/// Thrown instead of calling the API while the breaker is open.
struct CircuitOpenError: Error, Equatable {
    let retryAfter: TimeInterval
}

/// Thrown when the client-side throttle rejects a call.
struct RateLimitedError: Error, Equatable {}

/// Client-side circuit breaker + token-bucket throttle in one actor, so the
/// state is safe under Swift concurrency without locks.
///
/// A phone that keeps retrying a dead API drains the battery and delays the
/// moment the UI can honestly say "offline", so failing fast is a UX decision
/// here, not only an infrastructure one.
///
/// The guard deliberately does NOT run the request itself: handing a closure
/// across the actor boundary would drag Sendable requirements into every call
/// site. Callers ask for permission, do their own I/O, and report the outcome.
actor NetworkGuard {
    enum State: Equatable { case closed, open, halfOpen }

    private let failureThreshold: Int
    private let openDuration: TimeInterval
    private let perSecond: Double
    private let burst: Double
    private let now: @Sendable () -> Date

    private var state: State = .closed
    private var failures = 0
    private var openedAt: Date?
    private var tokens: Double
    private var lastRefill: Date?

    init(
        failureThreshold: Int = 4,
        openDuration: TimeInterval = 20,
        perSecond: Double = 8,
        burst: Double = 16,
        now: @escaping @Sendable () -> Date = Date.init
    ) {
        self.failureThreshold = failureThreshold
        self.openDuration = openDuration
        self.perSecond = perSecond
        self.burst = burst
        self.now = now
        self.tokens = burst
    }

    var currentState: State {
        refresh()
        return state
    }

    /// Throws when the circuit is open or the throttle is empty; returns
    /// normally when the caller may perform the request.
    func acquire() throws {
        refresh()
        if state == .open, let openedAt {
            throw CircuitOpenError(retryAfter: openDuration - now().timeIntervalSince(openedAt))
        }
        guard consumeToken() else { throw RateLimitedError() }
    }

    func recordSuccess() {
        state = .closed
        failures = 0
    }

    /// Only call this for a genuine dependency failure (transport error or
    /// 5xx). A 404 or a decoding bug says nothing about the backend's health.
    func recordFailure() {
        if state == .halfOpen {
            state = .open
            openedAt = now()
            return
        }
        failures += 1
        if failures >= failureThreshold {
            state = .open
            openedAt = now()
        }
    }

    private func consumeToken() -> Bool {
        guard perSecond > 0 else { return true }
        let current = now()
        let last = lastRefill ?? current
        tokens = min(burst, tokens + current.timeIntervalSince(last) * perSecond)
        lastRefill = current
        guard tokens >= 1 else { return false }
        tokens -= 1
        return true
    }

    private func refresh() {
        if state == .open, let openedAt, now().timeIntervalSince(openedAt) >= openDuration {
            state = .halfOpen
        }
    }
}
